File: Sources/QuillKit/Transcription/Vocabulary.swift
import Foundation

/// Words the recogniser has no reason to know.
///
/// A general model has never seen "graphify" or "Craigieburn", and a dictation
/// app that mangles the nouns you use every day is worse than useless — you
/// spend longer fixing it than typing would have taken. Apple's recogniser
/// accepts contextual strings as a bias, which is why this is fed to the
/// analyzer rather than fixed up afterwards: biasing changes what the model
/// *hears*, while find-and-replace can only repair what it already got wrong,
/// and cannot recover a word it split into two.
///
/// Verified need, not speculation: the first harness run transcribed "Netlify"
/// as "neglify" with no biasing in place.
public struct Vocabulary: Codable, Sendable, Equatable {

    public var terms: [String]

    /// Overridable, so a candidate dictionary can be scored against the corpus
    /// without being installed over the user's real one.
    ///
    /// Adding terms is not free: every entry is a standing chance to rewrite a
    /// word he meant, which is how "build a bed" became "Builda Bed". A change to
    /// this list has to be measured in both directions before it ships, and that
    /// is impossible if the only list the app can read is the live one.
    /// `QUILL_VOCABULARY_FILE` narrows further, to an exact file rather than a
    /// directory. Everything else, including plain `QUILL_DATA_DIR`, is
    /// `QuillData.directory` — the same source `QuillData.erase()` reads, so the
    /// two can never disagree about which file is real.
    public static let defaultURL: URL = {
        if let override = ProcessInfo.processInfo.environment["QUILL_VOCABULARY_FILE"], !override.isEmpty {
            return URL(fileURLWithPath: (override as NSString).expandingTildeInPath)
        }
        return QuillData.directory.appendingPathComponent("vocabulary.json")
    }()

    /// Nothing. A new install has an empty Dictionary.
    ///
    /// The list started as the author's life: his suburb, his school, his
    /// family, and eleven clients by full name. It shipped in the binary, so
    /// every stranger who installed Quill got a Dictionary of people they have
    /// never met — and those people never agreed to be in it. That was replaced
    /// with sixty-odd general developer terms, which was better and still
    /// wrong: the Dictionary is not a starter kit, it is the list of words THIS
    /// person says that a general model gets wrong, and nobody can guess that
    /// from outside.
    ///
    /// It is not cosmetic either. Every term is fed to the recogniser as
    /// biasing on every dictation and is what the vocabulary corrector matches
    /// against, so a shipped list is a standing instruction to hear words the
    /// user may never say.
    ///
    /// The empty state on the Dictionary screen does the job the seed was
    /// doing: it says what the feature is for, and offers the words Quill has
    /// already misheard — harvested from real dictations, so they are guesses
    /// about THEM.
    public static let seed = Vocabulary(terms: [])

    public init(terms: [String]) {
        self.terms = terms
    }

    public static func load(from url: URL = Vocabulary.defaultURL) -> Vocabulary {
        loadOutcome(from: url).vocabulary
    }

    /// The same read, with the one bit callers need before they write.
    ///
    /// `load` cannot distinguish "no file yet" from "a file I could not read",
    /// and both returned the shipped seed. Every writer does load → mutate →
    /// save, so one unreadable byte meant the next word added to the dictionary
    /// wrote thirty stock terms over everything the user had put there.
    public static func loadOutcome(from url: URL = Vocabulary.defaultURL)
        -> (vocabulary: Vocabulary, isDamaged: Bool) {
        switch StoreFile.read(Vocabulary.self, from: url) {
        case .missing:           return (seed, false)
        case .decoded(let read): return (read, false)
        case .unreadable:        return (seed, true)
        }
    }

    @discardableResult
    public func save(to url: URL = Vocabulary.defaultURL) -> Bool {
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.prettyPrinted, .sortedKeys]
        guard let data = try? encoder.encode(self) else { return false }
        try? FileManager.default.createDirectory(
            at: url.deletingLastPathComponent(), withIntermediateDirectories: true
        )
        return (try? data.write(to: url, options: .atomic)) != nil
    }

    /// Multi-word entries are kept whole. The API takes phrases, and splitting
    /// "Next Fulfilment" into two tokens would bias toward the common word
    /// "next" rather than the company.
    public var contextualStrings: [String] {
        terms
            .map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
            .filter { !$0.isEmpty }
    }
}

File: src/core/stores/vocabulary.ts
import { statSync } from 'node:fs';
import { dataFile, expandTilde, writeAtomic } from '../paths';
import { encodeJSON, readStoreFile } from './storeFile';
import { equalsIgnoringCase, trim } from '../text/strings';

/// Words the recogniser has no reason to know.
///
/// A general model has never seen "graphify" or "Craigieburn", and a dictation
/// app that mangles the nouns you use every day is worse than useless — you
/// spend longer fixing it than typing would have taken.
///
/// On macOS these were also handed to Apple's recogniser as contextual strings.
/// Measured, that did nothing: the same audio with 0 biasing terms and with 25
/// produced byte-identical text. Whisper does have a real biasing channel — the
/// `initial_prompt` — and the speech renderer uses it, but the correction pass
/// below is still what actually repairs a mangled name, because biasing is a
/// hint and not a guarantee.
export interface Vocabulary {
  terms: string[];
}

/// Seeded from the names a general speech model reliably mishears.
///
/// This list used to be the author's life: his suburb, his school, his family,
/// and eleven clients by full name. It shipped in the binary, so every stranger
/// who installed Quill got a Dictionary of people they have never met — and
/// those people never agreed to be in it. A seed is a guess at what ANY user
/// says; anything narrower belongs in the file on their own machine, which is
/// what the Dictionary tab is for.
export const VOCABULARY_SEED: Vocabulary = {
  terms: [
    'Firebase', 'Firestore', 'Netlify', 'Supabase', 'SQLite', 'Postgres',
    'Redis', 'Docker', 'Kubernetes', 'nginx', 'GraphQL', 'OAuth', 'JWT',
    'SwiftUI', 'SwiftPM', 'Xcode', 'TypeScript', 'JavaScript', 'Playwright',
    'PyTorch', 'NumPy', 'pandas', 'pytest', 'venv', 'CPython', 'codesign',
    'npm', 'pnpm', 'webpack', 'Vite', 'ESLint', 'Prettier', 'Tailwind',
    'React', 'Next.js', 'Node.js', 'Deno', 'Rust', 'Kotlin', 'Golang',
    'tmux', 'xterm', 'ssh', 'sudo', 'cron', 'regex', 'stdout', 'stderr',
    'API', 'CLI', 'SDK', 'UUID', 'JSON', 'YAML', 'CSV', 'HTTP', 'HTTPS',
    'CI', 'CD', 'repo', 'monorepo', 'changelog', 'hotfix', 'linting',
    // Cross-platform additions. The macOS build could assume a Mac vocabulary;
    // this one runs where people say these instead.
    'PowerShell', 'WSL', 'systemd', 'Wayland', 'Xorg', 'GNOME', 'KDE',
    'Ubuntu', 'Debian', 'Fedora', 'Arch', 'NixOS', 'winget', 'Chocolatey',
    'Electron', 'Vulkan', 'CUDA', 'ONNX', 'Whisper',
  ],
};

export function vocabularyURL(): string {
  const override = process.env.QUILL_VOCABULARY_FILE;
  if (override && override.length > 0) return expandTilde(override);
  return dataFile('vocabulary.json');
}

function validateVocabulary(value: unknown): Vocabulary | null {
  if (!value || typeof value !== 'object') return null;
  const terms = (value as { terms?: unknown }).terms;
  if (!Array.isArray(terms)) return null;
  if (!terms.every((term) => typeof term === 'string')) return null;
  return { terms: terms as string[] };
}

/// Multi-word entries are kept whole. Splitting "Next Fulfilment" into two
/// tokens would bias toward the common word "next" rather than the company.
export function contextualStrings(vocabulary: Vocabulary): string[] {
  return vocabulary.terms.map(trim).filter((term) => term.length > 0);
}

/// The same read, with the one bit callers need before they write.
///
/// A plain `load` cannot distinguish "no file yet" from "a file I could not
/// read", and both would return the shipped seed. Every writer does load →
/// mutate → save, so one unreadable byte would mean the next word added to the
/// dictionary writes sixty stock terms over everything the user had put there.
export function loadVocabularyOutcome(url = vocabularyURL()): {
  vocabulary: Vocabulary;
  isDamaged: boolean;
} {
  const outcome = readStoreFile(url, validateVocabulary);
  switch (outcome.kind) {
    case 'missing': return { vocabulary: { terms: [...VOCABULARY_SEED.terms] }, isDamaged: false };
    case 'decoded': return { vocabulary: outcome.value, isDamaged: false };
    case 'unreadable': return { vocabulary: { terms: [...VOCABULARY_SEED.terms] }, isDamaged: true };
  }
}

export function loadVocabulary(url = vocabularyURL()): Vocabulary {
  return loadVocabularyOutcome(url).vocabulary;
}

export function saveVocabulary(vocabulary: Vocabulary, url = vocabularyURL()): boolean {
  return writeAtomic(url, encodeJSON(vocabulary));
}

/// The vocabulary, live.
///
/// Reading the file once at launch meant a word added in the Dictionary went to
/// disk and then did nothing — not to the next dictation, not to the one after
/// that, only to the next launch. The screen said the word was in the
/// dictionary and the dictionary behaved as though it were not, which is the
/// worst of the three possible outcomes: a silent failure that looks like a
/// success.
///
/// So reads go through here. The file's modification time is checked on access
/// and the cache is refreshed when it moves. One `stat` per dictation is free
/// against a gap measured in seconds, and it means the file stays what its own
/// documentation promises — the source of truth, editable by hand, without the
/// app needing to be told.
export class VocabularyBook {
  private readonly url: string;
  private cached: Vocabulary;
  private stamp: number | null;

  private static sharedBook: VocabularyBook | null = null;
  static shared(): VocabularyBook {
    if (!VocabularyBook.sharedBook) VocabularyBook.sharedBook = new VocabularyBook();
    return VocabularyBook.sharedBook;
  }

  constructor(url: string = vocabularyURL()) {
    this.url = url;
    this.cached = loadVocabulary(url);
    this.stamp = VocabularyBook.modified(url);
  }

  /// The vocabulary as it is on disk right now.
  get current(): Vocabulary {
    const latest = VocabularyBook.modified(this.url);
    // Both null (no file, still no file) counts as unchanged, so a machine with
    // no vocabulary file does not re-read on every single call.
    if (latest !== this.stamp) {
      this.cached = loadVocabulary(this.url);
      this.stamp = latest;
    }
    return this.cached;
  }

  get terms(): string[] {
    return contextualStrings(this.current);
  }

  /// Adds a term, refusing duplicates and blanks. Returns false when nothing
  /// changed, so a caller can say "already there" rather than pretending.
  add(term: string): boolean {
    const trimmed = trim(term);
    if (trimmed.length === 0) return false;
    // Re-read before writing: the Dictionary screen and a hand edit can both be
    // in flight, and a stale in-memory copy written back would silently delete
    // whatever the other one added.
    const { vocabulary, isDamaged } = loadVocabularyOutcome(this.url);
    // Adding one word must never be the act that replaces the whole file with
    // the shipped seed.
    if (isDamaged) return false;
    if (contextualStrings(vocabulary).some((existing) => equalsIgnoringCase(existing, trimmed))) return false;
    vocabulary.terms.push(trimmed);
    if (!saveVocabulary(vocabulary, this.url)) return false;
    this.cached = vocabulary;
    this.stamp = VocabularyBook.modified(this.url);
    return true;
  }

  remove(term: string): boolean {
    const { vocabulary, isDamaged } = loadVocabularyOutcome(this.url);
    if (isDamaged) return false;
    const before = vocabulary.terms.length;
    vocabulary.terms = vocabulary.terms.filter((existing) => !equalsIgnoringCase(existing, term));
    if (vocabulary.terms.length === before) return false;
    if (!saveVocabulary(vocabulary, this.url)) return false;
    this.cached = vocabulary;
    this.stamp = VocabularyBook.modified(this.url);
    return true;
  }

  private static modified(url: string): number | null {
    try {
      return statSync(url).mtimeMs;
    } catch {
      return null;
    }
  }
}
